src/Npgsql/Util/StreamExtensions.cs

```diff
@@ -0,0 +1,98 @@
+﻿using System;
+using System.Buffers;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Npgsql.Util
+{
+    // Helpers to read/write Span/Memory<byte> to Stream before netstandard 2.1
+    internal static class StreamExtensions
+    {
+#if NET461 || NETSTANDARD2_0
+        public static int Read(this Stream stream, Span<byte> buffer)
+        {
+            var sharedBuffer = ArrayPool<byte>.Shared.Rent(buffer.Length);
+            try
+            {
+                var numRead = stream.Read(sharedBuffer, 0, buffer.Length);
+                if ((uint)numRead > (uint)buffer.Length)
+                {
+                    throw new IOException("Stream was too long.");
+                }
+                new Span<byte>(sharedBuffer, 0, numRead).CopyTo(buffer);
+                return numRead;
+            }
+            finally { ArrayPool<byte>.Shared.Return(sharedBuffer); }
+        }
+
+        public static ValueTask<int> ReadAsync(this Stream stream, Memory<byte> buffer, CancellationToken cancellationToken = default)
+        {
+            if (MemoryMarshal.TryGetArray(buffer, out ArraySegment<byte> array))
```

Comment:
This is a nice optimization, but remember that this is a backwards-compat shim for netstandard2.0/net461... Not sure the complexity of two paths here is warranted...

Reply:
Yes it is a nice one, but I have to be fair, I did not invent it 💡
It directly comes from [Stream.cs](https://github.com/dotnet/runtime/blob/622341ecc2d43ef81c509bde53de8513727d626d/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L453) 😋